### PR TITLE
[SERVER-1690][DataPoints] punch card

### PR DESCRIPTION
**Pull request by mksglu**

### CODE REVIEW

File: plugins/server-stats/frontend/public/javascripts/countly.models.js [hunk 1]
```diff
@@ -85,6 +112,447 @@
...
+
+        /**
+         * Convert hexadecimal to RGB value
+         * @param {string} hex value for color
+         * @returns {string} converted rgb value of hex color
+         */
+        function hexToRgb(hex) {
+            hex = hex.substr(1, 6);
+            var bigint = parseInt(hex, 16);
+            var r = (bigint >> 16) & 255;
```
Comment (ar2rsawseen) [inline plugins/server-stats/frontend/public/javascripts/countly.models.js]: Codacy Issue found: [Unexpected use of '&'.](https://app.codacy.com/manual/ar2rsawseen/countly-server/pullRequest?prid=5232873)

**Comment by ar2rsawseen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- plugins/server-stats/frontend/public/javascripts/countly.views.js  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ar2rsawseen/countly-server/pullRequest?prid=5232873&bid=17154661)

File: plugins/server-stats/frontend/public/javascripts/countly.models.js [hunk 2]
```diff
@@ -85,6 +112,447 @@
...
+            var g = (bigint >> 8) & 255;
```
Comment (ar2rsawseen) [inline plugins/server-stats/frontend/public/javascripts/countly.models.js]: Codacy Issue found: [Unexpected use of '&'.](https://app.codacy.com/manual/ar2rsawseen/countly-server/pullRequest?prid=5232873)

File: plugins/server-stats/frontend/public/javascripts/countly.models.js [hunk 3]
```diff
@@ -85,6 +112,447 @@
...
+            var b = bigint & 255;
```
Comment (ar2rsawseen) [inline plugins/server-stats/frontend/public/javascripts/countly.models.js]: Codacy Issue found: [Unexpected use of '&'.](https://app.codacy.com/manual/ar2rsawseen/countly-server/pullRequest?prid=5232873)

File: plugins/server-stats/api/api.js [hunk 4]
```diff
@@ -190,6 +190,62 @@ var plugins = require('../../pluginManager.js'),
...
+                            let mockMatrixColumn = dataPoints[property];
+                            const currentMatrixRow = days[property];
+                            for (let index = 0; index < mockMatrixColumn.length; index++) {
+                                const time = currentMatrixRow[index].dp;
+                                mockMatrixColumn[index] += time;
+                            }
+                        }
+                    }
+                }
+                if (error) {
```
Comment (ronkorving) [inline plugins/server-stats/api/api.js]: Wouldn't you want to reject early? If there's an error, there are likely no results so `results.length` on line 228 will crash hard.

File: plugins/server-stats/frontend/public/templates/data-points.html [hunk 5]
```diff
@@ -1,4 +1,4 @@
-<div class="widget">
+<div id="tod" class="widget">
```
Comment (ar2rsawseen) [inline plugins/server-stats/frontend/public/templates/data-points.html]: Would this indicate it requires styles for times-of-day?
if yes, then plugin should have its own styles defined in it's own css file
if no, let's rename it, not to clash with times of day :)

File: plugins/server-stats/api/api.js (hunk 4 above)
Comment (mksglu) [reply plugins/server-stats/api/api.js]: hi @ronkorving! thanks for your contribution. fixed.

**Comment by ar2rsawseen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- plugins/server-stats/frontend/public/javascripts/countly.views.js  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ar2rsawseen/countly-server/pullRequest?prid=5232873&bid=17154661)

File: plugins/server-stats/frontend/public/javascripts/countly.models.js (hunk 1 above)
Comment (ar2rsawseen) [inline plugins/server-stats/frontend/public/javascripts/countly.models.js]: Codacy Issue found: [Unexpected use of '&'.](https://app.codacy.com/manual/ar2rsawseen/countly-server/pullRequest?prid=5232873)

File: plugins/server-stats/frontend/public/javascripts/countly.models.js (hunk 2 above)
Comment (ar2rsawseen) [inline plugins/server-stats/frontend/public/javascripts/countly.models.js]: Codacy Issue found: [Unexpected use of '&'.](https://app.codacy.com/manual/ar2rsawseen/countly-server/pullRequest?prid=5232873)

File: plugins/server-stats/frontend/public/javascripts/countly.models.js (hunk 3 above)
Comment (ar2rsawseen) [inline plugins/server-stats/frontend/public/javascripts/countly.models.js]: Codacy Issue found: [Unexpected use of '&'.](https://app.codacy.com/manual/ar2rsawseen/countly-server/pullRequest?prid=5232873)

**Comment by ar2rsawseen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- plugins/server-stats/frontend/public/javascripts/countly.views.js  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ar2rsawseen/countly-server/pullRequest?prid=5232873&bid=17154661)

File: plugins/server-stats/frontend/public/javascripts/countly.models.js (hunk 1 above)
Comment (ar2rsawseen) [inline plugins/server-stats/frontend/public/javascripts/countly.models.js]: Codacy Issue found: [Unexpected use of '&'.](https://app.codacy.com/manual/ar2rsawseen/countly-server/pullRequest?prid=5232873)

File: plugins/server-stats/frontend/public/javascripts/countly.models.js (hunk 2 above)
Comment (ar2rsawseen) [inline plugins/server-stats/frontend/public/javascripts/countly.models.js]: Codacy Issue found: [Unexpected use of '&'.](https://app.codacy.com/manual/ar2rsawseen/countly-server/pullRequest?prid=5232873)

File: plugins/server-stats/frontend/public/javascripts/countly.models.js (hunk 3 above)
Comment (ar2rsawseen) [inline plugins/server-stats/frontend/public/javascripts/countly.models.js]: Codacy Issue found: [Unexpected use of '&'.](https://app.codacy.com/manual/ar2rsawseen/countly-server/pullRequest?prid=5232873)

**Pull request merged**